### ISSUE TITLE
Material property input order dependency

### ISSUE THREAD

**Issue by elementx54** [idaholab/moose#23508]
## Reason

I was creating and input file and found that if I moved the material block that declares my porosity, that I could avoid this error:
```
*** ERROR ***
Attempting to declare AD material property porosity, but it is already retrieved/declared as a regular material property.
```

I thought that MOOSE input files did not depend on the order of blocks. So, I went to see if I could get this same error using a MOOSE module test and I can. This is the patch for the tensor mechanics module in MOOSE.
[mech_porosity.patch](https://github.com/idaholab/moose/files/10798353/mech_porosity.patch)

After applying the patch, `./run_tests -j28 --re=lps_dual` will fail the test that was changed. The patch comments out a material model block and moves it to the end of the `[Material]` top block section. This shouldn't change anything.

This is just an example, the same thing happens in my input file and any other input that has `ADPorosityFromStrain` and `ADViscoplasticityStressUpdate` A lot of these test are found in `/modules/tensor_mechanics/test/tests/ad_viscoplasticity_stress_update` including the `lps_dual.i` that my patch changed. Try it out on other inputs and the same thing happens. Currently, these tests are passing due to the order in the input being "correct".

## Design

I don't really know what is going on, but I did notice that `ViscoplasticityStressUpdate` is getting an "Old" material property as a `<Real>`, even though it is and AD model. I'm guessing this tells something that the property is non-AD unless it is declared by the `type = ADPorosityFromStrain` model. The order of this "telling/declaring" depends on the order in the input file.

## Impact

I believe the use of the old material property is to avoid a cyclic dependency where porosity > strain > elasticity tensor > porosity.

**Comment by recuero** [comment]
Old material properties are supposed to be non-AD or regular as they do not have dependencies with the variables (contributions to the Jacobians are identically zero). We can take a look at that.

**Comment by recuero** [comment]
@dschwen it appears material property dependency resolution is failing in this case. Do you expect the fix you pasted in 
 https://github.com/idaholab/moose/issues/21793 would fix it?

**Comment by dschwen** [comment]
> @dschwen it appears material property dependency resolution is failing in this case. Do you expect the fix you pasted in #21793 would fix it?

I don't think so. That fix is for compute = false materials. I have encountered Adam's error as well before, but I think it was in connection with derivative properties. I only see an ADParsedMaterial here. I'll check if I can figure this out with the linked input.

**Comment by recuero** [comment]
It doesn't look to be `ADParsedMaterial` in this case. It's a porosity material property.

**Comment by dschwen** [comment]
ADPorosityFromStrain must come before ADViscoplasticityStressUpdate. If inserted right after it it fails already. Doesn't have to be all the way at the end.

**Comment by dschwen** [comment]
Ok, got it. `ADViscoplasticityStressUpdateBase` is _only_ getting the old porosity, and - as Antonio mentioned - old properties only exist as non-AD, because they do not depend on the current DOFs! MOOSE wrongly assumes that a never before seen property we are requesting the old state of, must be a non-AD property (while it could just as well be an AD property). 

There is not much we can do about this, other than adding a new disambiguating interface for getOldMaterialProperty, with an `is_ad` template parameter. I thought adding 

```
+    _porosity(this->template getGenericMaterialProperty<Real, is_ad>("porosity_name")),
```

would help (because it would MOOSE tell that the requested property is AD, but in this input that leads to a 

```
*** ERROR ***
Cyclic dependency detected in object ordering:
stress <- porosity <- stress
```

**Comment by dschwen** [comment]
Tag @lindsayad just FYI. I can create the disambiguating interface if you think that's a good way to avoid this problem.

**Comment by dschwen** [comment]
I suggest we simply add is_ad as a second template parameters that defaults to false. Then we start updating the framework and modules.

**Comment by lindsayad** [comment]
The new `getGenericMaterialPropertyOld<T, is_ad>` interface would require that the material class be templated right? What if there are multiple material properties that you're retrieving generically and some are non-AD and others are AD?

**Comment by dschwen** [comment]
> interface would require that the material class be templated right?

No, you can always just supply true or false.

**Comment by dschwen** [comment]
I'm not sure `getGenericMaterialPropertyOld` is the right name here, as the value that we're getting will always be non-AD. The `is_ad` parameter is more of a hint to MOOSE to correctly initialize the material data(?) object.

**Comment by lindsayad** [comment]
But my point is that generically speaking, you don't know whether the current state material property is non-AD or AD. Just as MOOSE got it wrong by "guessing" that the current state material property was going to be non-AD (and it turned out to be AD), with your new API (providing a template argument of `true`) you'd be explicitly indicating that the current state material property is going to be AD ... but what if the user in their input file actually specifies a material that creates the current state material property as non-AD?

**Comment by dschwen** [comment]
Well, currently a non-AD property will always be created. With the change developers could at least decide what flavor property is the right one. If the user provides the other flavor they are simply wrong :-D

**Comment by lindsayad** [comment]
Haha. Well your idea is definitely not bullet-proof but it is an improvement. I would welcome a PR that implements it

**Comment by dschwen** [comment]
Will do!

**Comment by paarkm** [comment]
I just ran into this yesterday and spent two hours debugging. The error message gives no indication that the order of the material blocks matters. Can we get this fixed for less experienced users?

**Comment by lindsayad** [comment]
This would be a good task for a BISON team member to work on using the MP TA charge number next FY
